chef/automate
# [automate-2629] add users for team workflow

### :nut_and_bolt: Description: What code changed, and why?

Here's the user story:
We want to add users to a particular team. We drill down to the team details, select "Add Users" and, lo and behold, all users that could be added have already been added, so there are no users to add.
But now there is a new message, with a big, enticing button to press.
<img width="961" alt="image" src="https://user-images.githubusercontent.com/6817500/72999407-2efd9600-3db4-11ea-9add-3fc39224c1be.png">

The goal of this PR is for that button to open the self-same user creation modal just recently updated for the users list page. Only now, the modal opens on top of the team's "Add User" dialog (left) -- and it is a dialog in a dialog (right).
<img width="1003" alt="image" src="https://user-images.githubusercontent.com/6817500/72999644-7f74f380-3db4-11ea-9e0e-673b163fcfa3.png">

Problem: that recently finished modal was not self-contained. No fault of that development--none of our existing creation/deletion modals are. They all require intricate wiring and handlers on the base page invoking them.
In order to avoid duplicating all that identical boilerplate, I am introducing a new design pattern: **the self-contained modal**. 

* Instead of 5 or 6 or 7 inputs and outputs, it needs just one: a signal to open and the modal manages itself. 
```
<app-create-user-modal [openEvent]="openUserModal"></app-create-user-modal>
```
* Instead of several methods, variables, and event emitters to manage communication between the modal and the base page, there is just one:
```
openModal(): void {
    this.openUserModal.emit(true);
  }
```
These improve (a) encapsulation and (b) adherence to the Single Responsibility Principle (the users' list should not need to know details about adding a user).

Once I had settled on the pattern implementation was very straightforward, thanks to the new modal (thanks @bcmdarroch and @blakestier !). Most of it was simply moving code from `UserFormComponent` and `UserManagementComponent` into the new `CreateUserModalComponent`. The rest was simplifying because we no longer need to manage the larger set of inputs/outputs to the modal.

### :chains: Related Resources
PR #2654 User modal update

### :+1: Definition of Done
You can add new local users from the teams "Add Users" page when the current list has been exhausted.

### :athletic_shoe: How to Build and Test the Change

Rebuild stencil library and automate-ui server:
```
cd components/automate-ui
make update-ui-lib
make serve
```

1. Navigate to Teams >> (team) >> Add Users.
2. If there are any users available to add, add them all, then re-open Add Users.
3. Select <kbd>Create User</kbd> and exercise the dialog. Do not forgot to try error conditions (attempting to add a user with an existing username, adding a password of "eeeeeeee", etc.)
4. For a successful addition, the dialog will close and the new user will appear in the list, ready to be added to the team.
4. Navigate to Users, select <kbd>Create User</kbd>, and exercise the dialog.

### :white_check_mark: Checklist
